apps/cowswap-frontend/src/modules/fiatAmount/apis/getCoingeckoPrice.ts

```diff
@@ -0,0 +1,60 @@
+import { SupportedChainId } from '@cowprotocol/cow-sdk'
+import { Token } from '@uniswap/sdk-core'
+
+import ms from 'ms.macro'
+
+export interface CoinGeckoUsdQuote {
+  [address: string]: {
+    usd: number
+  }
+}
+
+export const COINGECK_PLATFORMS: Record<SupportedChainId, string | null> = {
+  [SupportedChainId.MAINNET]: 'ethereum',
+  [SupportedChainId.GNOSIS_CHAIN]: 'xdai',
+  [SupportedChainId.GOERLI]: null,
+}
+
+const BASE_URL = 'https://api.coingecko.com/api/v3/simple/token_price'
+const VS_CURRENCY = 'usd'
+/**
+ * This is a text of 429 HTTP code
+ * https://saturncloud.io/blog/catching-javascript-fetch-failing-with-cloudflare-429-missing-cors-header/
+ */
+const FAILED_FETCH_ERROR = 'Failed to fetch'
+
+export const COINGECKO_RATE_LIMIT_TIMEOUT = ms`1m`
+
+export class CoingeckoRateLimitError extends Error {
+  constructor() {
+    super('CoingeckoRateLimitError')
+  }
+}
+
+export async function getCoingeckoPrice(currency: Token): Promise<number | null> {
+  const platform = COINGECK_PLATFORMS[currency.chainId as SupportedChainId]
+
+  if (!platform) throw new Error('UnsupporedCoingeckoPlatformError')
+
+  const params = {
+    contract_addresses: currency.address,
+    vs_currencies: VS_CURRENCY,
+  }
+
+  const url = `${BASE_URL}/${platform}?${new URLSearchParams(params)}`
+
+  return fetch(url)
```

Comment:
The module is used for two purposes:
1. Display fiat amounts
2. Display price impact and "price impact unknown" warning

Both those features are useful only immediately, especially "price impact unknown" warning.
We should display it as soon as possible, otherwise, it doesn't make sense, user just will make a swap without any warnings.
So, I think we shouldn't add back-off strategy here. I've added `CoingeckoRateLimitError` with a fallback to process 429 errors.

Reply:
My only concern is.

Imagine we have 3 requests per second in coingecko

- get USD in Conigecko for Token1
- get USD in Conigecko for Token2
- get USD in Conigecko for Token3
- get USD in CoW for Token4
- get USD in CoW for Token5

This might make our USD estimations behave a bit weird. 
If coingecko had a fair rate limit, we might want to reduce the speed at which we consume it. 
After all, this is for USD which normally we don't need too many at once, so if the rate limit is fair, it will not be noticeable difference.

So we can make the choice either:
- be as fast as possible and fallback right away --> PRO: faster, CON: can be inconsistent or hard to predict the results, plus failed request to CoW will result in no re-attempt
- Limit the speed to respect the limit of the APIs 

If fine either way, but i see we need to make a decision, and it was not for me super clear what is best in this case

Reply:
Added rate limiter and backOff:
https://github.com/cowprotocol/cowswap/pull/3081/commits/45b06154b2950f255ead8f824bc51a07464f2a80

Reply:
Note: `CoingeckoRateLimitError` is processed separately